### PR TITLE
feat: Add modules for encryption/decryption using Amazon's Key Management Service

### PR DESCRIPTION
**Pull request by ericrath**
**This PR includes 3 modules**
* `sda-commons-server-kms-aws`: Dropwizard bundle wrapping the aws-encryption-sdk for encryption and decryption using an AWS KMS
* `sda-commons-server-kms-aws-kafka`: wrappers for Serializer/Deserializer for encryption/decryption using `sda-commons-server-kms-aws`
* `sda-commons-server-kms-aws-testing`: AwsKms rule based on WireMock mocking an AWS KMS for encryption/decryption for integration test support.

**Documentation**
* Description of how to use and monitor AWS KMS: https://sda-se.atlassian.net/wiki/spaces/DG/pages/1907622003/Payload+Encryption
* Enhancement to the Developer Guide Kafka page: https://sda-se.atlassian.net/wiki/spaces/DG/pages/1915387924/Kafka+with+encryption

### CODE REVIEW

File: sda-commons-server-kms-aws/build.gradle [hunk 1]
```diff
@@ -0,0 +1,9 @@
+dependencies {
+  compile project(':sda-commons-server-starter')
+
+  compile "com.amazonaws:aws-java-sdk-kms"
+  compile "com.amazonaws:aws-java-sdk-sts"
+  compile "com.amazonaws:aws-encryption-sdk-java:2.0.0"
```
Comment (christopher-cudennec) [inline sda-commons-server-kms-aws/build.gradle:6]: I tend to move the version to `sda-commons-dependencies`

File: sda-commons-server-kms-aws/build.gradle [hunk 2]
```diff
@@ -1,5 +1,7 @@
 dependencies {
-  compile project(':sda-commons-server-starter')
+  compile project(':sda-commons-server-starter'), {
+    exclude group: 'io.openapitools.hal', module: 'swagger-hal'
+  }
```
Comment (Fox32) [inline sda-commons-server-kms-aws/build.gradle:4]: What is this dependency used for? I wondering a bit that something is depending on the old starter bundle

File: sda-commons-server-kms-aws/src/test/java/org/sdase/commons/server/kms/aws/util/MockKMSClient.java [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+      if (this == obj) return true;
+      if (obj == null) return false;
+      if (getClass() != obj.getClass()) return false;
+      DecryptMapKey other = (DecryptMapKey) obj;
+      if (cipherText == null) {
+        if (other.cipherText != null) return false;
+      } else if (!cipherText.equals(other.cipherText)) return false;
+      if (ec == null) {
+        if (other.ec != null) return false;
+      } else if (!ec.equals(other.ec)) return false;
```
Comment (Fox32) [inline sda-commons-server-kms-aws/src/test/java/org/sdase/commons/server/kms/aws/util/MockKMSClient.java:358]: Arg, please don't ever ever use if statements without code blocks. Adding a single line her can cause havoc. We even [have a rule for that](https://google.github.io/styleguide/javaguide.html#s4.1.1-braces-always-used). I guess the code can be simplified a bit.

File: sda-commons-server-kms-aws/src/test/java/org/sdase/commons/server/kms/aws/config/KmsAwsCachingConfigurationTest.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
+package org.sdase.commons.server.kms.aws.config;
+
+import static org.junit.Assert.*;
+
+import org.junit.Test;
+
+public class KmsAwsCachingConfigurationTest {
+  KmsAwsCachingConfiguration classUnderTest = new KmsAwsCachingConfiguration();
```
Comment (Fox32) [inline sda-commons-server-kms-aws/src/test/java/org/sdase/commons/server/kms/aws/config/KmsAwsCachingConfigurationTest.java:8]: Not sure I miss here something, but is `classUnderTest` a shared state between the tests? Are the tests order independent?

File: sda-commons-server-kms-aws/src/test/java/org/sdase/commons/server/kms/aws/config/KmsAwsConfigurationTest.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
+package org.sdase.commons.server.kms.aws.config;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+import org.junit.Test;
+
+public class KmsAwsConfigurationTest {
+  KmsAwsConfiguration classUnderTest = new KmsAwsConfiguration();
```
Comment (Fox32) [inline sda-commons-server-kms-aws/src/test/java/org/sdase/commons/server/kms/aws/config/KmsAwsConfigurationTest.java:9]: Not sure I miss here something, but is `classUnderTest` a shared state between the tests?

File: sda-commons-server-kms-aws/src/test/java/org/sdase/commons/server/kms/aws/dropwizard/AwsEncryptionTestApplication.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  @Override
+  public void run(AwsEncryptionTestConfiguration configuration, Environment environment) {
+    healthCheckRegistry = environment.healthChecks();
+  }
+
+  public KmsAwsBundle<AwsEncryptionTestConfiguration> AwsEncryptionBundle() {
+    return bundle;
+  }
+
+  public HealthCheckRegistry healthCheckRegistry() {
```
Comment (Fox32) [inline sda-commons-server-kms-aws/src/test/java/org/sdase/commons/server/kms/aws/dropwizard/AwsEncryptionTestApplication.java:34]: Is this used somewhere?

File: sda-commons-server-kms-aws/src/test/java/org/sdase/commons/server/kms/aws/KmsAwsPlainStringAndDecryptTest.java [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+
+    byte[] encrypt = kmsAwsEncryptionService.encrypt(encryptionInput.getBytes(), context);
+
+    KmsAwsDecryptionService kmsAwsDecryptionService =
+        new KmsAwsDecryptionServiceMock(false, kmsAwsEncryptionServiceMock.getMockKmsClient())
+            .getKmsAwsDecryptionService();
+
+    byte[] decrypt = kmsAwsDecryptionService.decrypt(encrypt, context);
+
+    Assert.assertEquals(encryptionInput, new String(decrypt));
```
Comment (Fox32) [inline sda-commons-server-kms-aws/src/test/java/org/sdase/commons/server/kms/aws/KmsAwsPlainStringAndDecryptTest.java:69]: We normally use [assertJ](https://joel-costigliola.github.io/assertj/) for assertions

File: sda-commons-server-kms-aws/build.gradle [hunk 8]
```diff
@@ -0,0 +1,19 @@
+dependencies {
+  compile project(':sda-commons-starter')
```
Comment (dhenneke) [inline sda-commons-server-kms-aws/build.gradle:2]: Modules shouldn't depend on the starter module but instead only on the direct dependencies that are actually need.

Comment (JoergSiebahn) [reply sda-commons-server-kms-aws/build.gradle:2]: I think, `sda-commons-server-dropwizard` is the dependency you are looking for.

File: sda-commons-server-kms-aws/src/test/java/org/sdase/commons/server/kms/aws/dropwizard/AwsEncryptionTestApplication.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+  @Override
+  public void initialize(Bootstrap<AwsEncryptionTestConfiguration> bootstrap) {
+    bootstrap.addBundle(ConfigurationSubstitutionBundle.builder().build());
+    bootstrap.addBundle(bundle);
+  }
+
+  @Override
+  public void run(AwsEncryptionTestConfiguration configuration, Environment environment) {
+    healthCheckRegistry = environment.healthChecks();
```
Comment (Fox32) [inline sda-commons-server-kms-aws/src/test/java/org/sdase/commons/server/kms/aws/dropwizard/AwsEncryptionTestApplication.java:27]: Do we need to assign this?

Comment (ericrath) [reply sda-commons-server-kms-aws/src/test/java/org/sdase/commons/server/kms/aws/dropwizard/AwsEncryptionTestApplication.java]: it's not necessary here. Removed it.

**Pull request closed without merging**